src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp

```diff
@@ -534,13 +534,31 @@ void os::verify_stack_alignment() {
 }
 #endif
 
+static THREAD_LOCAL WXMode _wx_state = WXUnknown;
 int os::extra_bang_size_in_bytes() {
   // AArch64 does not require the additional stack bang.
   return 0;
 }
 
-void os::current_thread_enable_wx(WXMode mode) {
-  pthread_jit_write_protect_np(mode == WXExec);
+WXMode os::current_thread_change_wx(WXMode new_state) {
+  if (_wx_state == WXUnknown) {
+    pthread_jit_write_protect_np(false /* not executable but writable */);
+    _wx_state = WXWrite;
+  }
+  WXMode old = _wx_state;
+  if (_wx_state != new_state) {
+    _wx_state = new_state;
+    pthread_jit_write_protect_np(_wx_state == WXExec);
+  }
```

Comment:
I would simplify this:

```
if (_wx_state == unknown) {
  _wx_state = write; // No way to know but we assume the original state is "writable, not executable"
}
WXMode old = _wx_state;
if (_wx_state != new_state) {
  _wx_state = new_state;
  pthread_jit_write_protect_np(_wx_state == WXExec);
}
```

that is simpler and avoids calling pthread_jit_write_protect_np twice for the "unknown->exec" transition.